From 298fda37c4c25434a15886ee9c261e566d595dff Mon Sep 17 00:00:00 2001
From: Aaron Kimball <aaron@cloudera.com>
Date: Fri, 18 Jun 2010 18:42:37 -0700
Subject: [PATCH 0313/1020] HADOOP-5203. TT's version build is too restrictive.

Description: Use the md5sum checksum of the source for determining version compatibility.
Reason: Improvement
Author: Rick Cox (0.20 backport by Bill Au)
Ref: CDH-1139
---
 .../org/apache/hadoop/HadoopVersionAnnotation.java |    7 +++++++
 src/core/org/apache/hadoop/util/VersionInfo.java   |   12 +++++++++++-
 src/saveVersion.sh                                 |    7 +++++--
 3 files changed, 23 insertions(+), 3 deletions(-)

diff --git a/src/core/org/apache/hadoop/HadoopVersionAnnotation.java b/src/core/org/apache/hadoop/HadoopVersionAnnotation.java
index 2bcd19f..04f0aea 100644
--- a/src/core/org/apache/hadoop/HadoopVersionAnnotation.java
+++ b/src/core/org/apache/hadoop/HadoopVersionAnnotation.java
@@ -53,4 +53,11 @@ public @interface HadoopVersionAnnotation {
    * @return the revision number as a string (eg. "451451")
    */
   String revision();
+
+  /**
+   * Get a checksum of the source files from which
+   * Hadoop was compiled.
+   * @return a string that uniquely identifies the source
+   **/
+  String srcChecksum();
 }
diff --git a/src/core/org/apache/hadoop/util/VersionInfo.java b/src/core/org/apache/hadoop/util/VersionInfo.java
index 6aee978..193ee46 100644
--- a/src/core/org/apache/hadoop/util/VersionInfo.java
+++ b/src/core/org/apache/hadoop/util/VersionInfo.java
@@ -79,6 +79,14 @@ public class VersionInfo {
   public static String getUrl() {
     return version != null ? version.url() : "Unknown";
   }
+
+  /**
+   * Get the checksum of the source files from which Hadoop was
+   * built.
+   **/
+  public static String getSrcChecksum() {
+    return version != null ? version.srcChecksum() : "Unknown";
+  }
   
   /**
    * Returns the buildVersion which includes version, 
@@ -88,12 +96,14 @@ public class VersionInfo {
     return VersionInfo.getVersion() + 
     " from " + VersionInfo.getRevision() +
     " by " + VersionInfo.getUser() + 
-    " on " + VersionInfo.getDate();
+    " source checksum " + VersionInfo.getSrcChecksum();
   }
   
   public static void main(String[] args) {
     System.out.println("Hadoop " + getVersion());
     System.out.println("Subversion " + getUrl() + " -r " + getRevision());
     System.out.println("Compiled by " + getUser() + " on " + getDate());
+    System.out.println("From source with checksum " + getSrcChecksum());
+
   }
 }
diff --git a/src/saveVersion.sh b/src/saveVersion.sh
index 22b7867..fb81bb6 100755
--- a/src/saveVersion.sh
+++ b/src/saveVersion.sh
@@ -29,15 +29,18 @@ fi
 url=`svn info | sed -n -e 's/URL: \(.*\)/\1/p'`
 user=`whoami`
 date=`date`
+srcChecksum=`find src -name '*.java' | LC_ALL=C sort | xargs md5sum | md5sum | cut -d ' ' -f 1`
+
 mkdir -p build/src/org/apache/hadoop
 cat << EOF | \
   sed -e "s/VERSION/$version/" -e "s/USER/$user/" -e "s/DATE/$date/" \
-      -e "s|URL|$url|" -e "s/REV/$revision/" \
+      -e "s|URL|$url|" -e "s/REV/$revision/" -e "s/SRCCHECKSUM/$srcChecksum/" \
       > build/src/org/apache/hadoop/package-info.java
 /*
  * Generated by src/saveVersion.sh
  */
 @HadoopVersionAnnotation(version="VERSION", revision="REV", 
-                         user="USER", date="DATE", url="URL")
+                         user="USER", date="DATE", url="URL",
+                         srcChecksum="SRCCHECKSUM")
 package org.apache.hadoop;
 EOF
-- 
1.7.0.4

